### PR TITLE
Squiz/ArrayDeclaration: fixes false positive when handling short lists inside `foreach`

### LINKED ISSUE #527: Array destructuring in foreach - conflict between SpaceBeforeComma & NoSpaceAfterComma
## Describe the bug
Destructuring an array for foreach - `foreach ($data as [, , $value]) {` - violates the `Squiz.Arrays.ArrayDeclaration.SpaceBeforeComma` ("Expected 0 spaces between "," and comma; 1 found")
But when "fixed" - `foreach ($data as [,, $value]) {` - the `Squiz.Arrays.ArrayDeclaration.NoSpaceAfterComma` is violated ("Expected 1 space between comma and ","; 0 found")
So this code cannot be written while using both sniffs together.

### Code sample
```php
class Foo
{
	/**
	 * @param list<array{int, int, int}> $data
	 */
	public function bar(array $data): int
	{
		$sum = 0;
		foreach ($data as [, , $value]) {
			$sum += $value;
		}

		return $sum;
	}
}
```

## Versions

|                                       |                                                                                                           |
|--------------------------|----------------------------------------------------------------------  |
| Operating System | Windows 11
| PHP version | 8.3
| PHP_CodeSniffer version | 3.10.1
| Install type | Composer

### PR DESCRIPTION
**Pull request by rodrigoprimo**
# Description

This PR fixes a false positive when handling a short list inside a `foreach` as reported in #527. I implemented the solution that @jrfnl described in this comment: https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/527#issuecomment-2189958063.

I included a separate commit with some tests that I wrote to improve test coverage for this sniff while I was familiarizing myself with its code. This is not a full code coverage commit. There are still some uncovered lines in this sniff, and those lines might be unreachable, but I have not investigated this at this time. There are also potentially some more relevant tests that could be added for lines that are already covered. This is also not addressed in this commit.

cc @VladaHejda in case you are available to test this PR and confirm if it fixes the problem that you reported. Thanks.

## Suggested changelog entry
`Squiz.Arrays.ArrayDeclaration`: fix false positive when handling short lists inside `foreach`

## Related issues/external references

Fixes #527

## Types of changes

- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

### CODE REVIEW

File: src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php [hunk 1]
```diff
@@ -45,6 +45,26 @@ public function process(File $phpcsFile, $stackPtr)
     {
         $tokens = $phpcsFile->getTokens();
 
+        // Bail when handling the most common cases of a short list inside a foreach (see
+        // https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/527).
+        if ($tokens[$stackPtr]['code'] === T_OPEN_SHORT_ARRAY
+            && isset($tokens[$stackPtr]['nested_parenthesis']) === true
+        ) {
+            $lastParenthesisCloser = end($tokens[$stackPtr]['nested_parenthesis']);
```
[the remark this answers is not in the document]
Comment (rodrigoprimo) [reply src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php]: Thanks! TIL!

**Comment by rodrigoprimo** [comment]
Thanks for your review, @jrfnl. I believe I replied to all your comments, and this PR is ready for another check.

> I appreciate the additional tests, but for the record: these should have been pulled in a separate PR as they have nothing to do with the issue you are addressing in this PR.

Noted. Please let me know if you prefer that I move the commit with the additional tests to a separate PR in this case.

File: src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php [hunk 2]
```diff
@@ -45,6 +45,26 @@ public function process(File $phpcsFile, $stackPtr)
...
+            $lastParenthesisOpener = key($tokens[$stackPtr]['nested_parenthesis']);
+            reset($tokens[$stackPtr]['nested_parenthesis']);
```
[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php:56]: Except that `foreach` acts on a _copy_ of the array and doesn't look at the internal array pointer at all, so sorry, that argument falls short.

[the remark this answers is not in the document]
Comment (rodrigoprimo) [reply src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php:56]: Now I can see my mistake. I tested what I said above last week using PsySH:

Screenshot from 2024-12-02 15-47-29

I accidentally pressed "Enter" between the `foreach` and the opening brace. What I didn't realize was that because of this, the `echo` statement was executed outside of the `foreach`. When I saw that only `5` was printed, I got to the incorrect conclusion that I mentioned above. Or at least, this is my current understanding of what happened.

Still, don't you think that it is a good practice to leave the array pointer pointing to the initial value? I believe other uses of the array will not use a copy and will use the internal array pointer. I can see that in other places in the PHPCS codebase where `end()` and `key()` are used, `reset()` is not called, so it is probably fine.

[the remark this answers is not in the document]
Comment (jrfnl) [reply src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php:56]: IIRC the internal array pointer is only used by the old-school `end()`, `key()`, `current()`, `prev()`, `next()`, `reset()` and `each()` functions.
Those should rarely be used in modern code.

Either way, to be honest, this whole discussion is academic as per commit c149b391d7a5add831886ad07972a96bff031556, `end()` is now acting on a _copy_ of the sub-array.

Though I noticed that commit c149b391d7a5add831886ad07972a96bff031556
 is incomplete and contains a bug now as `key()` is not using the sub-array, while it should...

In other words, the `$lastParenthesisOpener = key($tokens[$stackPtr]['nested_parenthesis']);` line needs to be fixed, then the `reset()` is 100% safe to remove.

[the remark this answers is not in the document]
Comment (rodrigoprimo) [reply src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php]: Thanks for the additional details, and I'm sorry I missed that the `$lastParenthesisOpener` also needed to be updated. This PR should be ready for a final review.

**Review by jrfnl — APPROVED** [review]
Thanks for finishing this off @rodrigoprimo, all good now.
